src/rdo.rs

```diff
@@ -364,39 +364,62 @@ pub fn rdo_tx_size_type<T: Pixel>(
   luma_mode: PredictionMode, ref_frames: [usize; 2], mvs: [MotionVector; 2], skip: bool
 ) -> (TxSize, TxType) {
   use crate::context::max_txsize_rect_lookup;
-  let tx_size = if !fi.tx_mode_select || !luma_mode.is_intra() {
-    max_txsize_rect_lookup[bsize as usize]
-  } else {
-    sub_tx_size_map[max_txsize_rect_lookup[bsize as usize] as usize]  // 1 level partition
-  };
-  debug_assert!(tx_size.width_log2() <= bsize.width_log2());
-  debug_assert!(tx_size.height_log2() <= bsize.height_log2());
 
-  // Luma plane transform type decision
+  let mut tx_size = max_txsize_rect_lookup[bsize as usize];
+  let mut best_tx_type = TxType::DCT_DCT;
+  let mut best_tx_size = tx_size;
+  let mut best_rd = std::f64::MAX;
   let is_inter = !luma_mode.is_intra();
-  let tx_set = get_tx_set(tx_size, is_inter, fi.use_reduced_tx_set);
+  let dct_type = [TxType::DCT_DCT];
 
-  let tx_type =
-    if tx_set > TxSet::TX_SET_DCTONLY && fi.config.speed_settings.rdo_tx_decision && !skip {
-      rdo_tx_type_decision(
-        fi,
-        fs,
-        cw,
-        luma_mode,
-        ref_frames,
-        mvs,
-        bsize,
-        bo,
-        tx_size,
-        tx_set,
-      )
-    } else {
-      TxType::DCT_DCT
-    };
+  let do_rdo_tx_size = fi.tx_mode_select && fi.config.speed_settings.rdo_tx_decision &&
+                luma_mode.is_intra();
+  let rdo_tx_depth = if do_rdo_tx_size { 1 } else { 0 };
+  let cw_checkpoint = cw.checkpoint();
+
+  for _ in 0..=rdo_tx_depth {
+    let tx_set = get_tx_set(tx_size, is_inter, fi.use_reduced_tx_set);
+
+    let do_rdo_tx_type = tx_set > TxSet::TX_SET_DCTONLY &&
+        fi.config.speed_settings.rdo_tx_decision && !skip;
 
-  assert!(tx_size.sqr() <= TxSize::TX_32X32 || tx_type == TxType::DCT_DCT);
+    if rdo_tx_depth == 0 && !do_rdo_tx_type { return (best_tx_size, best_tx_type) };
```

Comment:
@lu_zero, by 'early early', I guess you meant here?
It should happen at speed 3~10 (i.e. fi.config.speed_settings.rdo_tx_decision is false).
I see that the condition "rdo_tx_depth == 0 && !do_rdo_tx_type" seems a bit confusing.
It is equivalent to "!do_rdo_tx_size && !do_rdo_tx_type".